src/Psalm/Internal/Analyzer/Statements/Expression/SimpleTypeInferer.php

```diff
@@ -203,6 +203,61 @@ public static function infer(
             }
         }
 
+        if ($stmt instanceof PhpParser\Node\Expr\BitwiseNot) {
+            $stmt_expr_type = self::infer(
+                $codebase,
+                $nodes,
+                $stmt->expr,
+                $aliases,
+                $file_source,
+                $existing_class_constants,
+                $fq_classlike_name
+            );
+
+            if ($stmt_expr_type === null) {
+                return null;
+            }
+
+            $invalidTypes = clone $stmt_expr_type;
+            $invalidTypes->removeType('string');
+            $invalidTypes->removeType('int');
+            $invalidTypes->removeType('float');
+
+            if (!$invalidTypes->isUnionEmpty()) {
+                return null;
+            }
```

Comment:
I assume this causes stuff like `~[]` to infer as `mixed`? @orklah would we prefer to have it infer `int|string` since those are the only possible return values for `~`, even if it's an invalid operand?

Reply:
I don't really care one way or another but I guess mixed has the advantage of returning at least a warning at level 1 that may help with debugging